certbot-ci/certbot_integration_tests/utils/acme.py

```diff
@@ -0,0 +1,202 @@
+    acme_xdist = {'challtestsrv_port': 8055}
+
+    if acme_type == 'pebble':
+        acme_xdist['directory_url'] = 'https://localhost:14000/dir'
+    else:  # boulder
+        port = 4001 if acme_option == 'v2' else 4000
+        acme_xdist['directory_url'] = 'http://localhost:{0}/directory'.format(port)
+
+    acme_xdist['acme_subnet'] = '10.77.77' if acme_type == 'boulder' else '10.30.50'
+    acme_xdist['http_port'] = {node: port for (node, port)
+                               in zip(nodes, range(5200, 5200 + len(nodes)))}
+    acme_xdist['https_port'] = {node: port for (node, port)
+                                in zip(nodes, range(5100, 5100 + len(nodes)))}
+
+    return acme_xdist
+
+
+def _construct_workspace(acme_type):
+    """Create a temporary workspace for integration tests stack"""
+    workspace = tempfile.mkdtemp()
+
+    def cleanup():
+        """Cleanup function to call that will teardown relevant dockers and their configuration."""
+        for instance in [acme_type, 'traefik']:
+            print('=> Tear down the {0} instance...'.format(instance))
+            instance_path = join(workspace, instance)
+            try:
+                if os.path.isfile(join(instance_path, 'docker-compose.yml')):
+                    _launch_command(['docker-compose', 'down'], cwd=instance_path)
+            except subprocess.CalledProcessError:
+                pass
+            print('=> Finished tear down of {0} instance.'.format(acme_type))
+
+        shutil.rmtree(workspace)
+
+    # Here with atexit we ensure that clean function is called no matter what.
+    atexit.register(cleanup)
+
+    return workspace
+
+
+def _prepare_acme_server(workspace, acme_type, acme_option, acme_xdist):
+    """Configure and launch the ACME server, Boulder or Pebble"""
+    print('=> Starting {0} instance deployment...'.format(acme_type))
+    instance_path = join(workspace, acme_type)
+    try:
+        # This loads Boulder from git, that includes a docker-compose.yml ready for production.
+        if acme_type == 'boulder':
+            _launch_command(['git', 'clone', 'https://github.com/letsencrypt/boulder',
+                             '--single-branch', '--depth=1', instance_path])
+            # Allow Boulder to ignore usual limit rate policies, useful for tests.
+            os.rename(join(instance_path, 'test/rate-limit-policies-b.yml'),
+                      join(instance_path, 'test/rate-limit-policies.yml'))
+
+        # This configure Pebble using precompiled containers.
+        if acme_type == 'pebble':
+            os.mkdir(instance_path)
+            with open(join(instance_path, 'docker-compose.yml'), 'w') as file_h:
+                file_h.write('''\
+version: '3'
+services:
+  pebble:
+    image: letsencrypt/pebble
+    command: pebble -config /test/config/pebble-config.json {strict} -dnsserver 10.30.50.3:8053
+    ports:
+      - 14000:14000
+    networks:
+      acmenet:
+        ipv4_address: 10.30.50.2
```

Comment:
where are these addresses coming from?

Reply:
They are arbitrary adresses, in a similar way than Boulder Docker config. I took them from the original `docker-compose.yml` in Pebble repo, but their global value does not matter. What matter is that they form together a contained network where you can predictibly find pebble, challtestsrv and the Docker host. 

This is not standard from the Docker world. Normally you autodiscover services through their FQDN, that the Docker system maintain automatically, abstracting you the need to know the actual network architecture. But this is not possible at this moment, due to specifics in Boulder/Pebble, and the current Docker implementation for Linux (on Windows, a special FQDN is set for the Docker host).

Reply:
See also my comment below.